File: app/models/order.py
import uuid
from datetime import datetime
from app.extensions import db


class Order(db.Model):
    __tablename__ = 'orders'

    id = db.Column(db.Integer, primary_key=True)
    order_number = db.Column(db.String(50), unique=True, nullable=False, index=True)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=True)  # nullable for guest checkout
    guest_email = db.Column(db.String(255))
    guest_phone = db.Column(db.String(20))

    # Status
    status = db.Column(db.String(30), default='pending', nullable=False)
    # pending, confirmed, processing, shipped, delivered, cancelled, refunded

    # Pricing
    subtotal = db.Column(db.Numeric(10, 2), nullable=False)
    tax = db.Column(db.Numeric(10, 2), default=0)
    shipping_cost = db.Column(db.Numeric(10, 2), default=0)
    discount = db.Column(db.Numeric(10, 2), default=0)
    total = db.Column(db.Numeric(10, 2), nullable=False)

    # Addresses (stored as JSON string for order history immutability)
    shipping_address = db.Column(db.Text, nullable=False)
    billing_address = db.Column(db.Text)

    # Payment method: cod, online
    payment_method = db.Column(db.String(20), default='cod')

    # Coupon
    coupon_code = db.Column(db.String(50))

    # Notes
    notes = db.Column(db.Text)

    # Timestamps
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    shipped_at = db.Column(db.DateTime)
    delivered_at = db.Column(db.DateTime)

    # Relationships
    items = db.relationship('OrderItem', backref='order', lazy='dynamic', cascade='all, delete-orphan')
    payment = db.relationship('Payment', backref='order', uselist=False, cascade='all, delete-orphan')

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        if not self.order_number:
            self.order_number = f'SW-{uuid.uuid4().hex[:8].upper()}'

    @property
    def status_badge_class(self):
        status_map = {
            'pending': 'warning',
            'confirmed': 'info',
            'processing': 'primary',
            'shipped': 'info',
            'delivered': 'success',
            'cancelled': 'danger',
            'refunded': 'secondary',
        }
        return status_map.get(self.status, 'secondary')

    def __repr__(self):
        return f'<Order {self.order_number}>'


class OrderItem(db.Model):
    __tablename__ = 'order_items'

    id = db.Column(db.Integer, primary_key=True)
    order_id = db.Column(db.Integer, db.ForeignKey('orders.id'), nullable=False)
    variant_id = db.Column(db.Integer, db.ForeignKey('product_variants.id'))
    product_name = db.Column(db.String(300), nullable=False)
    product_slug = db.Column(db.String(350))
    size = db.Column(db.String(20))
    color = db.Column(db.String(50))
    price = db.Column(db.Numeric(10, 2), nullable=False)
    quantity = db.Column(db.Integer, nullable=False)
    image_url = db.Column(db.String(500))

    variant = db.relationship('ProductVariant', backref='order_items')

    @property
    def line_total(self):
        return self.price * self.quantity

    def __repr__(self):
        return f'<OrderItem {self.product_name} x{self.quantity}>'


class Payment(db.Model):
    __tablename__ = 'payments'

    id = db.Column(db.Integer, primary_key=True)
    order_id = db.Column(db.Integer, db.ForeignKey('orders.id'), nullable=False, unique=True)
    provider = db.Column(db.String(30), default='razorpay', nullable=False)
    razorpay_order_id = db.Column(db.String(255))
    razorpay_payment_id = db.Column(db.String(255))
    razorpay_signature = db.Column(db.String(500))
    amount = db.Column(db.Numeric(10, 2), nullable=False)
    currency = db.Column(db.String(10), default='INR')
    status = db.Column(db.String(30), default='pending')  # pending, captured, failed, refunded
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    def __repr__(self):
        return f'<Payment {self.razorpay_payment_id} - {self.status}>'

File: indistylex_app/lib/app/models/order.dart
class Order {
  final int id;
  final String orderNumber;
  final String status;
  final double total;
  final double subtotal;
  final double discount;
  final double shipping;
  final String? couponCode;
  final String paymentMethod;
  final String? paymentId;
  final String shippingAddress;
  final String? trackingNumber;
  final List<OrderItem> items;
  final DateTime createdAt;
  final DateTime? updatedAt;

  Order({
    required this.id,
    required this.orderNumber,
    required this.status,
    required this.total,
    this.subtotal = 0,
    this.discount = 0,
    this.shipping = 0,
    this.couponCode,
    this.paymentMethod = 'online',
    this.paymentId,
    required this.shippingAddress,
    this.trackingNumber,
    this.items = const [],
    required this.createdAt,
    this.updatedAt,
  });

  String get statusLabel {
    switch (status) {
      case 'pending':
        return 'Pending';
      case 'confirmed':
        return 'Confirmed';
      case 'processing':
        return 'Processing';
      case 'shipped':
        return 'Shipped';
      case 'delivered':
        return 'Delivered';
      case 'cancelled':
        return 'Cancelled';
      default:
        return status;
    }
  }

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['id'] ?? 0,
      orderNumber: json['order_number'] ?? '',
      status: json['status'] ?? 'pending',
      total: (json['total'] ?? 0).toDouble(),
      subtotal: (json['subtotal'] ?? 0).toDouble(),
      discount: (json['discount'] ?? 0).toDouble(),
      shipping: (json['shipping'] ?? 0).toDouble(),
      couponCode: json['coupon_code'],
      paymentMethod: json['payment_method'] ?? 'online',
      paymentId: json['payment_id'],
      shippingAddress: json['shipping_address'] ?? '',
      trackingNumber: json['tracking_number'],
      items: (json['items'] as List? ?? [])
          .map((item) => OrderItem.fromJson(item))
          .toList(),
      createdAt: DateTime.parse(
          json['created_at'] ?? DateTime.now().toIso8601String()),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
    );
  }
}

class OrderItem {
  final int id;
  final String productName;
  final String? productImage;
  final double price;
  final int quantity;
  final String? size;
  final String? color;

  OrderItem({
    required this.id,
    required this.productName,
    this.productImage,
    required this.price,
    required this.quantity,
    this.size,
    this.color,
  });

  double get total => price * quantity;

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      id: json['id'] ?? 0,
      productName: json['product_name'] ?? '',
      productImage: json['product_image'],
      price: (json['price'] ?? 0).toDouble(),
      quantity: json['quantity'] ?? 1,
      size: json['size'],
      color: json['color'],
    );
  }
}
